src/hooks/useFetchModels.ts

```diff
@@ -0,0 +1,103 @@
+import { useEffect, useReducer, Dispatch } from "react";
+
+import { createError } from "../utils/createError";
+
+export enum FetchModelsActionType {
+  FAILURE = "FAILURE",
+  SUCCESS = "SUCCESS",
+  FETCH_MODELS = "FETCH_MODELS",
+}
+
+export type FetchModelsAction<Model> =
+  | { type: FetchModelsActionType.FAILURE; payload: Error }
+  | { type: FetchModelsActionType.SUCCESS; payload: Model[] }
+  //The payload is whether to fetch models with different filter parameters or fetch more models with same filter parameters.
```

Comment:
I don't really understand this comment. The payload is a boolean, so... if it changes between true/false that indicates it needs to re-fetch?

Reply:
action = {type: "fetch_models", payload: true} means that it's request to fetch data -- i.e. replace `models` with the new list of `models` that is incoming.

action = {type: "fetch_models", payload: false} means it's a request to fetch more data -- i.e. append the incoming list of `models` to current list of `models`.

Basically, the first action is used when user did some filtering and this necessitate fetching data. The second action is using clicking "show more".

Reply:
action = {type: "fetch_models", payload: true} changes the state.fetchModels to true and state.fetchMoreModels to false

action = {type: "fetch_models", payload: false} changes state.fetchModels to false and state.fetchMoreModels to true

in the useEffect section, whenever state.fetchModels or state.fetchMoreModels changes, it will fetch data. Then it will dispatch an action of type `success` with payload=fetched models. Then depending on the value of state.fetchModels it replaces the models or append to it. https://github.com/tohuynh/cdp-frontend/blob/feature/paginate-votes/src/hooks/useFetchModels.ts#L34